source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc

```diff
@@ -0,0 +1,322 @@
+      PayloadToMetadata::NUMBER: {
+    double dval;
+    if (absl::SimpleAtod(StringUtil::trim(value), &dval)) {
+      val.set_number_value(dval);
+    } else {
+      ENVOY_LOG(debug, "value to number conversion failed");
+      return false;
+    }
+    break;
+  }
+  }
+
+  // Have we seen this namespace before?
+  auto namespace_iter = structs_by_namespace_.find(meta_namespace);
+  if (namespace_iter == structs_by_namespace_.end()) {
+    structs_by_namespace_[meta_namespace] = ProtobufWkt::Struct();
+    namespace_iter = structs_by_namespace_.find(meta_namespace);
+  }
+
+  auto& keyval = namespace_iter->second;
+  (*keyval.mutable_fields())[key] = val;
+
+  return true;
+}
+
+// add metadata['key']= value depending on payload present or missing case
+void PayloadToMetadataFilter::applyKeyValue(std::string value, const Rule& rule,
+                                            const KeyValuePair& keyval) {
+  if (keyval.has_regex_value_rewrite()) {
+    const auto& matcher = rule.regexRewrite();
+    value = matcher->replaceAll(value, rule.regexSubstitution());
+  } else if (!keyval.value().empty()) {
+    value = keyval.value();
+  }
+  // We do *not* modify value is not present in `on_present` or `on_missing`.
+  // That is, we apply the original field value to the metadata.
+
+  if (!value.empty()) {
+    const auto& nspace = decideNamespace(keyval.metadata_namespace());
+    addMetadata(nspace, keyval.key(), std::move(value), keyval.type());
+  } else {
+    ENVOY_LOG(debug, "value is empty, not adding metadata");
+  }
+}
+
+void PayloadToMetadataFilter::finalizeDynamicMetadata() {
+  if (!structs_by_namespace_.empty()) {
+    for (auto const& entry : structs_by_namespace_) {
+      decoder_callbacks_->streamInfo().setDynamicMetadata(entry.first, entry.second);
+    }
+  }
+}
+
+FilterStatus PayloadToMetadataFilter::messageBegin(MessageMetadataSharedPtr metadata) {
+  for (const auto& rule : config_->requestRules()) {
+    if (rule.matches(*metadata)) {
+      matched_rule_ids_.insert(rule.ruleId());
+    }
+  }
+
+  ENVOY_LOG(trace, "{} rules matched", matched_rule_ids_.size());
+  return FilterStatus::Continue;
+}
+
+FilterStatus PayloadToMetadataFilter::passthroughData(Buffer::Instance& data) {
+  if (!matched_rule_ids_.empty()) {
+    ASSERT(!decoder_);
+    handler_ = std::make_unique<TrieMatchHandler>(*this, config_->trieRoot());
+    transport_ = createTransport(decoder_callbacks_->downstreamTransportType());
+    protocol_ = createProtocol(decoder_callbacks_->downstreamProtocolType());
+    decoder_ = std::make_unique<Decoder>(*transport_, *protocol_, *handler_);
+
+    bool underflow = false;
+    decoder_->onData(data, underflow);
+    ASSERT(handler_->isComplete() || underflow);
+
+    finalizeDynamicMetadata();
+  }
+
+  return FilterStatus::Continue;
+}
+
+} // namespace PayloadToMetadataFilter
+} // namespace ThriftFilters
+} // namespace Extensions
+} // namespace Envoy
```

Comment:
is there any significant chunk of duplicate code in this file that could be shared with header-to-metadata? If so, let's move it to a common/ path (it can happen later too, and let's only do it if it's a non-trivial amount of lines).

Reply:
I put a TODO item here
https://github.com/envoyproxy/envoy/pull/23409/files#diff-7c9ea70a9bcf34c08c4213b100d808ad3c5462ba16624cbc37dace202cd87621R172

The most apparent part to extract is `applyKeyValue` and `addMetadata` 

The reasons I did not reuse are:
(a) the implementation replies on the proto so it would need either template or some additional proto->enum mapping
https://github.com/envoyproxy/envoy/blob/6bba2d69a17e4cf374eb09f7f51fbf0f0ce4b7da/source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc#L100

and 
https://github.com/envoyproxy/envoy/blob/6bba2d69a17e4cf374eb09f7f51fbf0f0ce4b7da/source/extensions/filters/network/thrift_proxy/filters/header_to_metadata/header_to_metadata_filter.cc#L140

where `KeyValuePair` is tied to proto. 

(b) h2m `setDynamicMetadata` once but p2m does it while we parse the tree.

It's definitely doable but the refactor is non-trivial.
Hence, I'd like to do the refactor once the filter is stable in proto and production.